api/grpc_server.go

```diff
@@ -59,32 +62,44 @@ func (s SpacemeshGrpcService) GetNonce(ctx context.Context, in *pb.AccountId) (*
 
 	val := s.StateApi.GetNonce(addr)
 	msg := pb.SimpleMessage{Value: strconv.FormatUint(val, 10)}
-	log.Error("GRPC GetNonce returned msg %v", strconv.FormatUint(val, 10))
+	log.Info("GRPC GetNonce returned msg %v", strconv.FormatUint(val, 10))
 	return &msg, nil
 }
 
 func (s SpacemeshGrpcService) SubmitTransaction(ctx context.Context, in *pb.SignedTransaction) (*pb.SimpleMessage, error) {
 	log.Info("GRPC SubmitTransaction msg")
+
+	signedTx := &types.SerializableSignedTransaction{}
+	rdr := bytes.NewReader(in.Tx)
+	_, err := xdr.Unmarshal(rdr, signedTx)
+	if err != nil {
+		log.Error("failed to deserialize tx, error %v", err)
+		return nil, err
+	}
+
+	src, err := s.Tx.ValidateTransactionSignature(*signedTx)
+	if err != nil {
+		log.Error("tx failed to validate signature, error %v", err)
+		return nil, err
+	}
+
+	// TODO once the node will support signed txs we should only use types.SerializableSignedTransaction instead of types.SerializableTransaction
 	tx := types.SerializableTransaction{}
-	addr := address.HexToAddress(in.DstAddress)
-	tx.Recipient = &addr
-	tx.Origin = address.HexToAddress(in.SrcAddress)
+	tx.Recipient = &signedTx.Recipient
+	tx.Origin = src
 
-	num, _ := strconv.ParseInt(in.Nonce, 10, 64)
-	tx.AccountNonce = uint64(num)
+	tx.AccountNonce = signedTx.AccountNonce
 	amount := big.Int{}
-	amount.SetString(in.Amount, 10)
+	amount.SetBytes(signedTx.Amount)
 	tx.Amount = amount.Bytes()
-	tx.GasLimit = 10
-	tx.Price = big.NewInt(10).Bytes()
+	tx.GasLimit = signedTx.GasLimit
+	tx.Price = signedTx.Price
 
 	val, err := types.TransactionAsBytes(&tx)
 	if err != nil {
 		return nil, err
 	}
-
-	//todo" should this be in a go routine?
-	s.Network.Broadcast(miner.IncomingTxProtocol, val)
+	go s.Network.Broadcast(miner.IncomingTxProtocol, val)
```

Comment:
can you use the received tx buffer to avoid encoding again? also, what about the tx signature? should it be re signed before broadcasting?

Reply:
The struct transferred on the wire is different than the struct that is processed by the node. Once the node will support tx's signature we can have the same struct for both uses.

Reply:
add todo here

Reply:
there is already a TODO (line 83)